apps/api/src/http/v3/rest/egvault/redeem.rs

```diff
@@ -0,0 +1,216 @@
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use axum::extract::State;
+use axum::{Extension, Json};
+use shared::database::entitlement::{EntitlementEdge, EntitlementEdgeId, EntitlementEdgeKind, EntitlementEdgeManagedBy};
+use shared::database::product::codes::{CodeEffect, RedeemCode};
+use shared::database::product::subscription::SubscriptionId;
+use shared::database::product::TimePeriod;
+use shared::database::queries::{filter, update};
+use shared::database::user::UserId;
+
+use super::{create_checkout_session_params, find_customer};
+use crate::global::Global;
+use crate::http::error::ApiError;
+use crate::http::middleware::auth::AuthSession;
+use crate::transactions::{with_transaction, TransactionError, TransactionResult, TransactionSession};
+
+pub async fn grant_entitlements(
+	tx: &mut TransactionSession<'_, ApiError>,
+	redeem_code: &RedeemCode,
+	user_id: UserId,
+) -> TransactionResult<(), ApiError> {
+	for (to, extends_subscription) in redeem_code.effects.iter().filter_map(|e| match e {
+		CodeEffect::Entitlement {
+			edge,
+			extends_subscription,
+		} => Some((edge, extends_subscription)),
+		_ => None,
+	}) {
+		let from = match extends_subscription {
+			Some(extends_subscription) => EntitlementEdgeKind::Subscription {
+				subscription_id: SubscriptionId {
+					user_id: user_id,
+					product_id: *extends_subscription,
+				},
+			},
+			None => EntitlementEdgeKind::User { user_id },
+		};
+
+		tx.insert_one(
+			EntitlementEdge {
+				id: EntitlementEdgeId {
+					from,
+					to: to.clone(),
+					managed_by: Some(EntitlementEdgeManagedBy::RedeemCode {
+						redeem_code_id: redeem_code.id,
+					}),
+				},
+			},
+			None,
+		)
+		.await?;
+	}
+
+	Ok(())
+}
```

Comment:
Does this run at all if there is a subscription grant in the effects? I think we should delay this to run only after the sub has been completed or if the user is already subscribed to the product.

Reply:
That's exactly how it works right now